### PR TITLE
[WIP] AWS metadata processor

### LINKED ISSUE #926: Dynamically set global tags
It would be nice to have a way to dynamically set the global tags.
Some use cases would be
- add some portions of ec2-metadata
- add AWS tags as global tags

Some of them may change over time (especially AWS tags) so putting them in the config file is not ideal.

### PR DESCRIPTION
**Pull request by glinton**
closes #926

Continuation of #2036

### CODE REVIEW

File: plugins/processors/aws/ec2/ec2.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+		Region:    e.Region,
+		AccessKey: e.AccessKey,
+		SecretKey: e.SecretKey,
+		RoleARN:   e.RoleARN,
+		Profile:   e.Profile,
+		Filename:  e.Filename,
+		Token:     e.Token,
+	}
+	configProvider := credentialConfig.Credentials()
+	// e.client = ec2.New(configProvider)
```
Comment (danielnelson) [inline plugins/processors/aws/ec2/ec2.go]: Remove commented code

File: plugins/processors/aws/sqs/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# SQS Metadata Processor Plugin
+
+The SQS Metadata processor plugin appends additional metadata from AWS to metrics associated with Simple Queue Service.
+
+### Configuration:
+
+```toml
+## Annotate metrics from the cloudwatch plugin
+[[processors.aws_metadata_sqs]]
```
Comment (danielnelson) [inline plugins/processors/aws/sqs/README.md]: Indent

File: plugins/processors/aws/elb/elb.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	internalaws "github.com/influxdata/telegraf/internal/config/aws"
+	"github.com/influxdata/telegraf/plugins/processors"
+	"github.com/influxdata/telegraf/plugins/processors/aws/utils"
```
Comment (emanuelr93) [inline plugins/processors/aws/elb/elb.go]: Seems that this package missing in commit. Can you check?

**Comment by naumfujimori** [comment]
Guys any update regarding to this feature?
Regrds,

**Comment by naumfujimori** [comment]
Guys,
Any update regarding to this feature?

**Comment by danielmotaleite** [comment]
So around 2 years have gone by since the original commit, one year after the comments, was this abandoned?

Could it be simpler to just call ec2metadata and grab the data? it is more limited, but should be enough for many people and requires no authentication

**Comment by ssoroka** [comment]
This stalled originally because of a lack of support for asynchronous processors. Since then, we have added support for asynchronous processors (StreamingProcessor) 🎉, which now makes supporting this possible. 
Adding it to the plan for 1.16 as 1.15 is imminent.

**Comment by ssoroka** [comment]
If anyone's really keen on tackling this before we get to it, you'll want to check out the [reverse_dns processor](https://github.com/influxdata/telegraf/blob/master/plugins/processors/reverse_dns/reversedns.go#L86-L142) which does something very similar, using the [StreamingProcessor](https://github.com/influxdata/telegraf/blob/master/processor.go#L13-L31).

**Comment by pmalek-sumo** [comment]
@ssoroka I've submitted a draft PR https://github.com/influxdata/telegraf/pull/8707 that focuses only on AWS EC2 metadata using `StreamingProcessor`.

Please comment if that's something that is acceptable. One thing (besides general comments) I'd like to hear is whether configurable mapping of tags (e.g. `instanceId` -> `instance-id`) would be a wanted feature.

**Comment by sheldonhull** [comment]
Wouldn't the string/replacement processors (don't remember the exact one) handle any renaming of the tags?

I'm thankful for the continued effort on this. My manual bootstrapping of this in the config is much more brittle. Using the datadog agent recently made me appreciate just how important the ec2 tagging metadata is to simplify metric analysis.

**Comment by pmalek-sumo** [comment]
> Wouldn't the string/replacement processors (don't remember the exact one) handle any renaming of the tags?

I didn't think about that! Yes, it would (you just have to specify the order between the processors):

```toml
[[processors.aws_ec2]]
  order = 1
  tags = [ "accountId" ]

[[processors.rename]]
  order = 2
  [[processors.rename.replace]]
    tag = "accountId"
    dest = "account-id"
```

**Comment by pmalek-sumo** [comment]
Since https://github.com/influxdata/telegraf/pull/8707 got merged shouldn't this PR be closed? Or is there an intention to use other AWS processors from this PR?

**Comment by ssoroka** [comment]
resolved in https://github.com/influxdata/telegraf/pull/8707

**Pull request closed without merging**